daemon/src/peer/run_state/running_waiting_room.rs

```diff
@@ -0,0 +1,345 @@
+            cmds.push(Command::Request(command::Request::Query(urn)));
+            cmds.push(Command::PersistWaitingRoom(self.waiting_room.clone()));
+        }
+        if let Some((urn, remote_peer)) = self.waiting_room.next_clone() {
+            cmds.push(Command::Request(command::Request::Clone(urn, remote_peer)));
+            cmds.push(Command::PersistWaitingRoom(self.waiting_room.clone()));
+        }
+
+        let state_after = self.waiting_room.requests();
+        let transition = WaitingRoomTransition {
+            timestamp: now,
+            state_before,
+            state_after,
+            event: Event::Tick,
+        };
+        cmds.push(Command::EmitEvent(transition.into()));
+        cmds
+    }
+
+    pub fn cloning(
+        &mut self,
+        urn: &Urn,
+        remote_peer: PeerId,
+        timestamp: SystemTime,
+    ) -> Vec<Command> {
+        Self::simple_command_helper(
+            &mut self.waiting_room,
+            urn,
+            Event::Cloning {
+                urn: urn.clone(),
+                peer: remote_peer,
+            },
+            timestamp,
+            |w| w.cloning(urn, remote_peer, timestamp),
+        )
+    }
+
+    pub fn cloned(
+        &mut self,
+        urn: &Urn,
+        remote_peer: PeerId,
+        timestamp: SystemTime,
+    ) -> Vec<Command> {
+        Self::simple_command_helper(
+            &mut self.waiting_room,
+            urn,
+            Event::Cloned {
+                urn: urn.clone(),
+                peer: remote_peer,
+            },
+            timestamp,
+            |w| w.cloned(urn, remote_peer, timestamp),
+        )
+    }
+
+    pub fn queried(&mut self, urn: &Urn, timestamp: SystemTime) -> Vec<Command> {
+        Self::simple_command_helper(
+            &mut self.waiting_room,
+            urn,
+            Event::Queried { urn: urn.clone() },
+            timestamp,
+            |w| w.queried(urn, timestamp),
+        )
+    }
+
+    pub fn cloning_failed(
+        &mut self,
+        urn: &Urn,
+        remote_peer: PeerId,
+        timestamp: SystemTime,
+        reason: Box<dyn std::error::Error>,
+    ) -> Vec<Command> {
+        Self::simple_command_helper(
+            &mut self.waiting_room,
+            urn,
+            Event::CloningFailed {
+                urn: urn.clone(),
+                peer: remote_peer,
+                reason: reason.to_string(),
+            },
+            timestamp,
+            |w| w.cloning_failed(urn, remote_peer, timestamp, reason),
+        )
+    }
+
+    fn simple_command_helper<F>(
+        waiting_room: &mut WaitingRoom<SystemTime, Duration>,
+        urn: &Urn,
+        event: Event,
+        timestamp: SystemTime,
+        f: F,
+    ) -> Vec<Command>
+    where
+        F: FnOnce(&mut WaitingRoom<SystemTime, Duration>) -> Result<(), WaitingRoomError>,
+    {
+        let state_before = waiting_room.requests();
+        // FIXME(alexjg): Come up with a strategy for the results returned by the
```

Comment:
What do you mean here?

Reply:
This is actually a copy of a `FIXME` @xla had in `run_state.rs`. I've assumed responsibility for it because I'm  touching this code and not fixing it. The basic problem is that we don't actually report the error in this result to the client, we just log it and carry on, it seems like we should probably have some API for telling the client that something has gone wrong.

Reply:
Ahhh that makes sense. Could you maybe change the wording to mention reporting the error back? Or even move the comment to the error case.

Reply:
Done